### PR TITLE
WIP: Fix redirects

### PR DESCRIPTION
**Pull request by maraoz**

### CODE REVIEW

**Comment by maraoz** [comment]
This is working but I still have to fix a problem with a totally unrelated test. I seem to have caused a regression or uncovered a previous bug. See my other comment

File: test/index.ts [hunk 1]
```diff
@@ -890,7 +941,6 @@ if (!popsicle.browser) {
     t.test('should follow 303 redirect with post', function (t) {
       return popsicle.post({
         url: REMOTE_URL + '/redirect/code/303',
-        body: { foo: 'bar' }
```
Comment (maraoz) [inline test/index.ts]: Had to remove this line for the test to work. With the line present, the test hangs, and I have no idea why. Any ideas?

**Comment by maraoz** [comment]
btw, your test environment is a nightmare! console.logs dont work!!! :-1:

**Comment by maraoz** [comment]
I opened the incomplete PR for review and to get help on the regression, btw :)

**Comment by blakeembrey** [comment]
> btw, your test environment is a nightmare! console.logs dont work!!! :-1:

Sadly, I know :frowning: To make them come out, you should remove the `| faucet`. I've moved in newer projects to using a better formatter that does output `console.log` on failing tests.

> I opened the incomplete PR for review and to get help on the regression, btw :)

This is amazing, thanks! I'll review it tonight and I might pick some code from here. How was playing with TypeScript, at least?

**Comment by blakeembrey** [comment]
@maraoz Found the regression and refactored a bit of the code. I'll make another PR with it, but the issue was the redirect did not update the `Content-Length` header when the body was removed so the server was hanging waiting for more data. Interesting it wasn't caught before this, but I just made a change to kill the Content-Length header on a GET redirect where the body is removed.

**Comment by blakeembrey** [comment]
Thanks so much for this! I just merged a separate attempt with some improvements. I think the only thing that was missing from here was the patch mentioned above, and also remembering to remove old cookies before we redirect.

**Pull request closed without merging**